source/elements/oneDAL/source/data_management/array.rst

```diff
@@ -99,39 +99,83 @@ usage scenario:
+3. If an array represents immutable data, pointer to the mutable data block shall be ``nullptr``.
 
-5. An array shall store a pointer to the **ownership structure** of the data:
+4. An array shall use shared ownership semantics to manage lifetime of the stored data blocks:
 
-   - The **reference count** indicating how many array objects refer to the
-     same memory block.
+   - Several arrays objects may own the same data block;
 
-   - The **deleter** object used to free the memory block when
-     reference count is zero.
+   - The memory block is deallocated when either of the following happens:
 
-6. An array shall create the ownership structure for a new memory block not
-   associated with such structure.
+     - The last remaining array owning data block is destroyed;
 
-7. An array shall decrement the number of references to the memory block when the
-   array goes out of the scope. If the number of references is zero, the
-   array shall call the deleter on this memory block and free the ownership structure.
+     - The last remaining array owning data block is assigned another memory block via
+       ``operator=`` or ``reset()``;
+
+   - The data block is deallocated using the deleter object that is provided to array during
+     construction.
+
+5. An array object may own no data. In this case, it is called **zero-sized**:
+
+   - Pointers to the immutable and mutable data of the zero-sized array shall be ``nullptr``;
+   - The data block size shall be ``0``.
+
+
+.. _implementation_notes:
+
+--------------------
+Implementation notes
+--------------------
+A typical array implementation may be organized in the following way:
+
+1. An array class has the following member variables:
+
+   - Pointer to the immutable data block;
+
+   - Pointer to the mutable data block;
+
+   - Pointer to the ownership structure that implements the shared ownership semantics;
+
+   - The data block size;
+
+2. The ownership structure is an object that stores:
+
+   - Either a pointer immutable or mutable data block;
+
+   - The deleter object;
+
+   - The reference count (the number of array instances that owns the associated data block);
+
+3. The destructor of an array decrements the reference count. If that count reaches zero, the
+   ownership structure deallocates the associated memory block and the array destroys the ownership
+   structure.
+
+4. If the array object changes its state from immutable to mutable, the reference count of the
+   ownership structure is decremented.
+
+   - If that count reaches zero, the ownership structure
+     deallocates the immutable memory block and the array destroys the ownership structure. The new
+     instance of ownership structure owning the mutable data block is created and replaces the old one.
+
+   - If that count greater than zero, the ownership structure is not destroyed. The new
+     instance of ownership structure owning the mutable data block is created and replaces the old
```

Comment:
```suggestion
     instance of the ownership structure owning the mutable data block is created, and it replaces the old
```

if you repeat the same sentence for both cases, shouldn't it be just a single sentence?